Dynamic Dashboard: Load data for stock card

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12796 
Closes #12754 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR completes the functionality for the stock dashboard card with logic and content on the UI. Changes include:
- Updated the view model to load data. The logic is to load stock report based on the selected stock type, then either fetch product reports & product details or use in-memory data if already loaded before.
- Mapped loaded details to a new model `StockItem` to display on the stock card UI.
- Handled empty state and error states for the card.
- Handled navigation from the stock item to open the detail screen for the relevant product.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to the app and enable the Stock card.
- Switch between different stock type and confirm that the card displays the correct items of the selected status with the correct stock quantity and number of items sold in the last 30 days. To confirm the data, you can use Core's Analytics to cross-check.
- Tap on any stock item, confirm that the detail form is displayed for that product.
- Go to wp-admin > WooCommerce > Settings > Advanced > Features > disable Woo Analytics and reload the app.
- Confirm that the error state for disabled Analytics is displayed.
- Enable Analytics again, then disconnect the Internet before reloading. Confirm that the generic error state screen is displayed.
- Enable the connection then switch to a store with no stock of a certain status. Confirm that the empty state is displayed when that status is selected.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
Empty state | Analytics disabled | Generic error | With data
--- | --- | --- | ---
<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/b138237a-978c-4925-b87c-fad1243398c4" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/bc379df9-0718-4f6a-b50d-f4b51a4ac44c" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/ddd32652-7a6a-4468-bed5-23c73caba487" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/7151d8b8-3d87-4132-af36-398ba7cb9f7d" width=320 /> 



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add remote requests for stock card

itsmeichigo committed:
Fix invalid request params

itsmeichigo committed:
Handle syncing data for stock card

itsmeichigo committed:
Display stock details on stock card

itsmeichigo committed:
Add empty state for product card

itsmeichigo committed:
Update width for empty state view

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Add error states for stock card

itsmeichigo committed:
Remove storage from stock card

itsmeichigo committed:
Update tests for ProductStockDashboardCardViewModel

itsmeichigo committed:
Show product detail upon tapping stock items

itsmeichigo committed:
Sort items by stock quantity

itsmeichigo commented:
One issue I'm seeing: whenever switching stock status, the card's order is pushed to the bottom. I don't think this was caused by any logic inside the card but more likely to happen from the DashboardViewModel. This will need to be handled in a separate PR.

selanthiraiyan commented:
Hey @itsmeichigo 👋 

I did an initial round of testing using Jurassic Ninja store with products and orders generated using https://github.com/woocommerce/wc-smooth-generator. 

### On back order

I see differences in data between mobile and core. Is this because we don't consider product variations? 

Please let me know if I have missed any discussions about this. 

| Core | Mobile |
|--------|--------|
| <img width="1094" alt="Screenshot 2024-05-22 at 7 38 26 AM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/c4238855-ba51-4f17-8d8d-d88c83cdd753"> | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-05-22 at 07 38 30](https://github.com/woocommerce/woocommerce-ios/assets/524475/6cc8700c-b6c3-46ba-8fb7-e0df247f5641) | 

### Items sold in the last 30 days

There is a difference in the number of items sold as well. 

| Core | Mobile |
|--------|--------|
| <img width="1091" alt="Screenshot 2024-05-22 at 7 39 58 AM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/783bff26-cd9c-4e85-9e91-59da31c719b9"> | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-05-22 at 07 38 30](https://github.com/woocommerce/woocommerce-ios/assets/524475/f003a343-fbad-4201-8e90-15814f693d5d) |

itsmeichigo committed:
Update item order for product stock request

itsmeichigo committed:
Fix incorrect before and after date for product report request

itsmeichigo commented:
Thanks @selanthiraiyan for the review!

- The difference in items was due to the order used in the stock request. I updated the order following the web app in 66608c8b00e5a7bc59ae2cbbe5b0d61e7315248e.
- The incorrect items sold was due to incorrect dates used for the date range in the product report request. This has been fixed in c2d45f0a45b2711eed557a10254b80aabf17a229.

This PR is ready for another look.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCard.swift:
```diff
@@ -154,6 +250,47 @@ private extension ProductStockDashboardCard {
+            static let done = NSLocalizedString(
+                "productStockDashboardCard.contentUnavailable.dismissSupport",
+                value: "Done",
+                comment: "Button to dismiss the support form from the Dashboard stats error screen."
```
nit
```suggestion
                comment: "Button to dismiss the support form from the Dashboard stock card error screen."
```

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/ProductStockDashboardCardViewModelTests.swift:
```diff
@@ -1,4 +1,5 @@
+import Yosemite
 @testable import WooCommerce
 
 final class ProductStockDashboardCardViewModelTests: XCTestCase {
```
Nice tests. ❤️

selanthiraiyan reviewed (commented):
Thanks for the updates, Huong!

For the "Low stock" option I see that the items sold in last 30 days value is not loaded.
<img width="320" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/731f686a-de76-4e78-87b2-da2eb75e334d"/>

I looked into the network requests, and it looks like the response is empty while trying to load those products using `wc/v3/products` endpoint.
<img width="1256" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/a94e9bf3-c128-45be-8a9a-b9761385d8d0">

Meanwhile I see that those products have orders in core. 
<img width="1099" alt="Screenshot 2024-05-22 at 1 03 59 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/acd1fed2-80e7-4ab2-8772-aa79b00a86c2">

1. Is this because those products are variations? 
1. Can we avoid showing the "0 items sold last 30 days" text if we don't have the information for a particular product?

itsmeichigo committed:
Update WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCard.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/204">18.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented:
Thanks @selanthiraiyan for testing! The above issue is indeed due to variations, we'll need to add extra support for that case. I'll set this PR back to draft to handle that case.

> Can we avoid showing the "0 items sold last 30 days" text if we don't have the information for a particular product?

Any suggestion? Should we show "No item sold last 30 days" or just hide the text completely?

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Restore content for package.resolved

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Update logic using ProductReport

itsmeichigo committed:
Fix unit tests

itsmeichigo committed:
Fetch product and variation reports

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Display more friendly text for zero items sold

itsmeichigo committed:
Update parent ID for variation reports with invalid product ID

itsmeichigo committed:
Display product variation when selected

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Remove orderby for stock report request

itsmeichigo committed:
Fix unit test build failure

itsmeichigo committed:
Update test to include fetching variation report fetching

itsmeichigo commented:
Hey @selanthiraiyan 👋 Sorry for a lot of new commits - but I think the PR is ready for another look. Changes since your last review include:
- Removed `orderby` param when fetching stock reports as it's redundant.
- Added support for fetching product variation reports. 
- Removed product detail requests since the items sold count and thumbnail URLs are available in the reports.
- Updated navigation from stock item to open product variation detail screen for variations.
- Updated the wording to "No item sold last 30 days" for items not sold within the period.
- Always fetch product reports after stock reports to have fresh numbers of items sold. We can introduce an expiration period for the cache to optimize performance later if necessary.

itsmeichigo committed:
Always fetch new reports to have fresh number of items sold

selanthiraiyan commented on WooCommerce.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -84,7 +84,7 @@
       {
         "package": "swift-numerics",
-        "repositoryURL": "https://github.com/apple/swift-numerics",
+        "repositoryURL": "https://github.com/apple/swift-numerics.git",
```
❓Just to confirm. Is this an accidental commit?

itsmeichigo replied:
I accidentally pushed a diff of the file that removed all the dependencies when rebasing the branch, and then I tried to revert it but ended up with this change. I'm not sure how to revert this, but I hope this is not a big problem because CI can still resolve the dependencies with this. WDYT?

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCard.swift:
```diff
@@ -23,15 +28,43 @@ struct ProductStockDashboardCard: View {
+                if viewModel.reports.isNotEmpty {
+                    stockList
+                } else {
+                    emptyView
```
Suggestion: When the card is loaded for the first time, we end up showing the empty view in redacted mode. I think we can avoid this by showing `stockList` also when `syncingData` is true. WDYT?

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/312be136-a68d-4159-8d6e-e1501a2f7c3f" width="320"/>

itsmeichigo replied:
Good point! Updated in f7f3da85c4065cc36bbcd3166dc38cbf2cd29b1d. The card should now look like this:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/f6de9a24-fe16-488e-b410-64ae009dc3c9" width=320 />

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift:
```diff
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+    @MainActor
+    func fetchStock(type: StockType) async throws -> [ProductStock] {
+        try await withCheckedThrowingContinuation { continuation in
+            /// `order` and `orderby` follow the same setting on Woo Analytics page
```
❓ We don't specify `orderby`, right? 🤔

itsmeichigo replied:
Removed the outdated comment in 11acae02c18ea01ad931c8924297d6e8e838c330.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift:
```diff
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+            stores.dispatch(ProductAction.fetchProductReports(siteID: siteID,
+                                                              productIDs: productIDs,
+                                                              timeZone: timeZone,
+                                                              earliestDateToInclude: currentDate,
+                                                              latestDateToInclude: last30Days,
```
nit: I think `earliestDateToInclude` should carry the start date, which in context is `last30Days` value. `latestDateToInclude` should carry the end date value which is `currentDate`. We might also have to update the remote [here](https://github.com/woocommerce/woocommerce-ios/blob/69a2d3d167661280099db42e6991cbac1ab62a8b/Networking/Networking/Remote/ProductsRemote.swift#L472-L473).

The same comment applies to the `fetchVariationReports` method as well. 

What do you think?

itsmeichigo replied:
Good catch! I fixed this in 12c5cf4c8f73d2ca6e892ff91dca58aa66280f79.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift:
```diff
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+    @MainActor
+    func fetchAndSaveReportsToMemory(for stock: [ProductStock]) async throws {
+
+        let groupedStockByVariations = Dictionary(grouping: stock, by: { $0.isProductVariation })
+        let variationsToFetchReports = groupedStockByVariations[true] ?? []
+        let productsToFetchReports = groupedStockByVariations[false] ?? []
```
Nice. 👏

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift:
```diff
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+            if variationsToFetchReports.isNotEmpty {
+                group.addTask {
+                    let reports = try await self.fetchVariationReports(
+                        productIDs: Array(Set(variationsToFetchReports.map { $0.parentID })),
+                        variationIDs: variationsToFetchReports.map { $0.productID }
```
❓I believe that variation ID is not unique across products. Please correct me if I am wrong about this. 

Considering that the above is true, when we are fetching reports for more than one variation, there could be a conflict in variation IDs. I wonder how the endpoint is able to differentiate variations with the same IDs but they actually belong to different product IDs. 🤔

itsmeichigo replied:
I confirmed with the team in p1716794405419699-slack-C03L1NF1EA3 that variation IDs are globally unique so this solution should work as expected.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift:
```diff
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+                }
+            }
+
+            while !group.isEmpty {
+                // rethrow any failure.
+                try await group.next()
+            }
```
❓ For my learning: Could you help me understand the purpose of this code?

itsmeichigo replied:
I learned this while working on the Performance cards where I had to trigger several network requests in a task group. Without using `try await group.next()`, each task would fail silently. Using this makes sure that the group can rethrow any error occurring in its child tasks. You can find more details in the documentation [here](https://developer.apple.com/documentation/swift/taskgroup/next()).

selanthiraiyan reviewed (approved):
Not a blocker: To get the dashboard work as expected (for a fresh login), I had to bring in changes from `trunk` to get this PR https://github.com/woocommerce/woocommerce-ios/pull/12826 code in. 

---

### "Low Stock" variations

The order and stock levels are not the same when comparing the mobile app and core. Please feel free to merge if this is an issue with the backend. 

<img width="1430" alt="Screenshot 2024-05-27 at 8 38 16 AM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/fca6c2b2-9477-4e72-bda1-af201ec4da92">

--- 

Thank you for the updates and your patience, Huong! 🙇 

I left a couple of nits/comments. I left a question about how we are fetching variation's reports. 

🚀

itsmeichigo committed:
Merge branch 'feat/12796-fetch-variation-report' into feat/12796-stock-card-logic

itsmeichigo committed:
Remove outdated comment

itsmeichigo committed:
Do not shimmer empty view when syncing data in the empty state

itsmeichigo committed:
Fix mixed up date ranges for product and variation report requests

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews! I responded to your comments, please check them out when you have time.

Regarding this:

> The order and stock levels are not the same when comparing the mobile app and core. Please feel free to merge if this is an issue with the backend.

I believe from your screenshot that you were ordering the items by stock count. The web also fetches more items than us, that's why their response includes an item with stock count of 2. Does that make sense?

selanthiraiyan commented:
Thanks for the updates, Huong! This looks great now. 🚀 

> I believe from your screenshot that you were ordering the items by stock count. The web also fetches more items than us, that's why their response includes an item with stock count of 2. Does that make sense?

Yes. I understand. 👍
